### ISSUE TITLE
alps::mcmpiadapter::run() blocking communication leads to deadlocks

### ISSUE THREAD

**Issue by aeantipov** [ALPSCore/ALPSCore#116]
Got a following problem in `alps::mc::mcmpiadapter`
When I run `mcmpiadapter::run()`, it calls `MPI_AllReduce` to update the progress of calculation after every `update()` and `measure()`. Normally this doesn't cause any problems. However, when the run-time of `update()` operation varies substantially between processors and the interval between subsequent calls is smaller than the maximum `update()` run-time, this results in other processes being deadlocked. 
Right now when I profile the code I see 50% of time being spent on `MPI_AllReduce`. It can be alleviated by increasing the checking interval time, but that doesn't completely solve it. This problem is somewhat fixed in new mpi standard by providing non-blocking collective communication, but backwards-compatibility likely requires to solve it. In the end, it's a single double variable to be updated.

**Comment by galexv** [comment]
And this looks like a classical load balancing problem. It can be solved, e.g., by using an (emulated) shared memory counter (like GAMESS does). I will look at it...

**Comment by galexv** [comment]
This is postponed to Release 0.5 for now (meantime, the increase of the checking interval time can be used as a workaround). Just to clarify the severity: were actual deadlocks (never-ending waits) observed, or it just spends too much time on waiting?

**Comment by aeantipov** [comment]
The second - it spends too much time waiting for the slowest of the processes to finish it's task

**Comment by egull** [comment]
Removing the bug label from this one. I think it behaves as designed, but you have a special situation and we'll need to see how we deal with it.

**Comment by galexv** [comment]
Is it still an actual problem --- are there ALPSCore-using codes out there that suffer from significant load imbalance? I propose to put this issue into "icebox" for now.

**Comment by egull** [comment]
I don't think we know. We would have to test this thoroughly to see if this problem ever appears. I agree with moving to icebox.
